contracts/contracts/libraries/PolynomialEval.sol

```diff
@@ -180,18 +182,38 @@ library PolynomialEval {
         //      = Z_H(zeta) * v_i / (zeta - g^i)
         //      = vanish_eval_div_n * g^i / (zeta - g^i)
         // - v_i = g^i / n
+        divisorProd = 1;
         for (uint256 i = 0; i < length; i++) {
             assembly {
                 // tmp points to g^i
                 // first 32 bytes of reference is the length of an array
                 tmp := mload(add(add(localDomainElements, 0x20), mul(i, 0x20)))
-                // vanish_eval_div_n * g^i
-                ithLagrange := mulmod(vanishEvalDivN, tmp, p)
                 // compute (zeta - g^i)
                 divisor := addmod(sub(p, tmp), zeta, p)
             }
+            // compute prod (zeta - g^i)
+            buffer[i] = divisor;
+            divisorProd = mulmod(divisorProd, divisor, p);
+        }
+
+        // compute 1 / prod (zeta - g^i)
+        divisorProd = BN254.invert(divisorProd);
+
+        for (uint256 i = 0; i < length; i++) {
+            assembly {
+                // tmp points to g^i
+                // first 32 bytes of reference is the length of an array
+                tmp := mload(add(add(localDomainElements, 0x20), mul(i, 0x20)))
+                // vanish_eval_div_n * g^i
+                ithLagrange := mulmod(vanishEvalDivN, tmp, p)
+            }
             // compute 1/(zeta - g^i)
-            divisor = BN254.invert(divisor);
+            divisor = divisorProd;
+            for (uint256 j = 0; j < length; j++) {
+                if (i != j) {
+                    divisor = mulmod(divisor, buffer[j], p);
+                }
+            }
```

Comment:
we can turn this into full assembly easily so that we can have a single assembly block in this for-loop, wdyt (you can decide after comparing the gas cost, I agree that right now your code is more readable)?

```solidity
ithDivsiorInv := divisiorProd
for { let j := 0 } lt(j, length) { i := add(i, 1) }
{
    if not(eq(i, j)) { 
        ithDivisor := mload(add(buffer, add(0x20, mul(j, 0x20))))
        ithDivisorInv := mulmod(ithDivisorInv, ithDivisor, p) 
    }
}
```

Reply:
good idea. we can put more comment in the code and use your version